added Twig_Source to hold information about the original template

fabpot opened the pull request:
Alternative to #2170

nicolas-grekas commented on lib/Twig/Node/Module.php:
```diff
@@ -40,8 +47,11 @@ public function __construct(Twig_NodeInterface $body, Twig_Node_Expression $pare
-            'source' => $source,
-            'filename' => $filename,
+            'source' => $source->getCode(),
+            // to be renamed in 2.0 to name
```
then let's add name right now?

nicolas-grekas commented on lib/Twig/Node/Module.php:
```diff
@@ -40,8 +47,11 @@ public function __construct(Twig_NodeInterface $body, Twig_Node_Expression $pare
+            'source' => $source->getCode(),
+            // to be renamed in 2.0 to name
+            'filename' => $source->getName(),
+            // to be renamed in 2.0 to filename
```
but that would be a hard break, no continuous upgrade path

stof commented on lib/Twig/Loader/Filesystem.php:
```diff
@@ -14,7 +14,7 @@
  * @author Fabien Potencier <fabien@symfony.com>
  */
-class Twig_Loader_Filesystem implements Twig_LoaderInterface, Twig_ExistsLoaderInterface
+class Twig_Loader_Filesystem implements Twig_LoaderInterface, Twig_ExistsLoaderInterface, Twig_SourceContextLoaderInterface
```
the chain loader should implement Twig_SourceContextLoaderInterface too, otherwise we loose the capabilities when using multiple loaders

fabpot replied:
done

stof commented on lib/Twig/Template.php:
```diff
@@ -65,6 +65,16 @@ public function getSource()
+     */
+    public function getSourceContext()
+    {
+        return new Twig_Source('');
```
``$this->getTemplateName()`` should be passed as second argument here

stof commented on lib/Twig/Source.php:
```diff
@@ -0,0 +1,44 @@
+    private $name;
+    private $filename;
+
+    public function __construct($code, $name = null, $filename = null)
```
please add phpdoc to document the variable types

fabpot replied:
fixed

stof commented on lib/Twig/SourceContextLoaderInterface.php:
```diff
@@ -0,0 +1,18 @@
+
+interface Twig_SourceContextLoaderInterface
+{
+    /**
```
missing phpdoc for the argument

fabpot replied:
fixed

stof commented on lib/Twig/TokenStream.php:
```diff
@@ -32,9 +32,19 @@ class Twig_TokenStream
+            if (null !== $filename || null !== $source) {
+                @trigger_error(sprintf('Passing a string as the $filename argument of %s() is deprecated since version 1.27. Pass a Twig_Source instance instead.', __METHOD__), E_USER_DEPRECATED);
+            }
+            $this->source = new Twig_Source($filename, $source);
```
this is wrong. the first argument of ``Twig_Source`` is the code

stof replied:
Please also add legacy tests covering the deprecated APIs. They would catch such mistakes (and forbid breaking BC by mistake)

fabpot replied:
fixed, tests coming

stof commented on lib/Twig/TokenParser/AutoEscape.php:
```diff
@@ -39,7 +39,7 @@ public function parse(Twig_Token $token)
             $expr = $this->parser->getExpressionParser()->parseExpression();
             if (!$expr instanceof Twig_Node_Expression_Constant) {
-                throw new Twig_Error_Syntax('An escaping strategy must be a string or a bool.', $stream->getCurrent()->getLine(), $stream->getFilename());
+                throw new Twig_Error_Syntax('An escaping strategy must be a string or a bool.', $stream->getCurrent()->getLine(), $this->parser->getFilename());
```
why not using ``$stream->getSourceContext()->getName()`` directly to avoid 1 level in the stack each time ? the parser will forward the call to the stream anyway

fabpot commented:
Rebased on top of #2183, reviewing the second commit only gives you less noise. Renamed `getFilename()` to get `getPath()`. I think it's much better that way.

fabpot commented:
Tests added, not WIP anymore

stof commented on lib/Twig/Loader/Chain.php:
```diff
@@ -66,6 +66,35 @@ public function getSource($name)
+        $exceptions = array();
+        foreach ($this->loaders as $loader) {
+            if (!$loader instanceof Twig_SourceContextLoaderInterface) {
+                continue;
```
this is broken in case this loader was the one loading the template, but a next loader could load a template with the same name

fabpot replied:
fixed in #2192

marc1706 commented:
Causes a BC break when upgrading from 1.26 and prior to twig 1.27 as described here:
https://github.com/twigphp/Twig/commit/6c76583e23537b2503ccd555d053c09a8250b8ac#commitcomment-19686237

xabbuh commented:
@stof But that doesn't mean that someone could have implemented the interface before which will no indeed break their code when updating to Twig 1.27.

marc1706 commented:
My reply to @stof 's comment:
https://github.com/twigphp/Twig/commit/6c76583e23537b2503ccd555d053c09a8250b8ac#commitcomment-19693147

Should probably continue the discussion here as that's more accessible to other users than having the discussion in the commit.
